controlnet-1.5-face/predict.py

```diff
@@ -0,0 +1,322 @@
+            unet=self.txt2img_pipe.unet,
+            scheduler=self.txt2img_pipe.scheduler,
+            safety_checker=self.txt2img_pipe.safety_checker,
+            feature_extractor=self.txt2img_pipe.feature_extractor,
+        )
+
+        print("Loading controlnet...")
+        controlnet = ControlNetModel.from_pretrained(
+            settings.CONTROLNET_MODEL,
+            subfolder="diffusion_sd15",
+            torch_dtype=torch.float16,
+            cache_dir=settings.MODEL_CACHE,
+            local_files_only=True,
+        )
+
+        print("Loading controlnet txt2img...")
+        self.cnet_txt2img_pipe = StableDiffusionControlNetPipeline(
+            vae=self.txt2img_pipe.vae,
+            text_encoder=self.txt2img_pipe.text_encoder,
+            tokenizer=self.txt2img_pipe.tokenizer,
+            unet=self.txt2img_pipe.unet,
+            scheduler=self.txt2img_pipe.scheduler,
+            safety_checker=self.txt2img_pipe.safety_checker,
+            feature_extractor=self.txt2img_pipe.feature_extractor,
+            controlnet=controlnet,
+        ).to("cuda")
+
+        print("Loading controlnet img2img...")
+        self.cnet_img2img_pipe = StableDiffusionControlNetImg2ImgPipeline(
+            vae=self.txt2img_pipe.vae,
+            text_encoder=self.txt2img_pipe.text_encoder,
+            tokenizer=self.txt2img_pipe.tokenizer,
+            unet=self.txt2img_pipe.unet,
+            scheduler=self.txt2img_pipe.scheduler,
+            safety_checker=self.txt2img_pipe.safety_checker,
+            feature_extractor=self.txt2img_pipe.feature_extractor,
+            controlnet=controlnet,
+        )
+
+        print("Loading inpaint...")
+        self.inpainting_pipe = StableDiffusionInpaintPipelineLegacy(
+            vae=self.txt2img_pipe.vae,
+            text_encoder=self.txt2img_pipe.text_encoder,
+            tokenizer=self.txt2img_pipe.tokenizer,
+            unet=self.txt2img_pipe.unet,
+            scheduler=self.txt2img_pipe.scheduler,
+            safety_checker=self.txt2img_pipe.safety_checker,
+            feature_extractor=self.txt2img_pipe.feature_extractor,
+        )
+
+        print("Loading compel...")
+        self.compel = Compel(
+            tokenizer=self.txt2img_pipe.tokenizer,
+            text_encoder=self.txt2img_pipe.text_encoder,
+        )
+
+        self.real = True
+
+    def load_image(self, image_path: Path):
+        if image_path is None:
+            return None
+        # not sure why I have to copy the image, but it fails otherwise
+        # seems like a bug in cog
+        if os.path.exists("img.png"):
+            os.unlink("img.png")
+        shutil.copy(image_path, "img.png")
+        return load_image("img.png")
+
+    def process_control(self, control_image):
+        if control_image is None:
+            return None
+
+        # TODO do some actual pre-processing instead of returning the already pre-processed control image
+        return control_image
```

Comment:
This currently **just returns** the already preprocessed control image. We could add some actual preprocessing here, eg. by implementing https://huggingface.co/CrucibleAI/ControlNetMediaPipeFace/blob/main/tool_generate_face_poses.py